flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/FsStateChangelogWriter.java

```diff
@@ -0,0 +1,207 @@
+    private final NavigableMap<SequenceNumber, StateChangeSet> changeSets = new TreeMap<>();
+    private final NavigableMap<SequenceNumber, StoreResult> uploaded = new TreeMap<>();
+    private final NavigableMap<SequenceNumber, StoreResult> confirmed = new TreeMap<>();
+    private List<StateChange> activeChangeSet = new ArrayList<>();
+    private SequenceNumber lastAppendedSequenceNumber = SequenceNumber.of(0L);
+    private boolean closed;
+    private final ChangelogCallbackExecutor executor;
+
+    FsStateChangelogWriter(
+            UUID logId,
+            KeyGroupRange keyGroupRange,
+            StateChangeStore store,
+            ChangelogCallbackExecutor executor) {
+        this.logId = logId;
+        this.keyGroupRange = keyGroupRange;
+        this.store = store;
+        this.executor = Preconditions.checkNotNull(executor);
+    }
+
+    @Override
+    public void append(int keyGroup, byte[] value) {
+        LOG.trace("append to {}: keyGroup={} {} bytes", logId, keyGroup, value.length);
+        checkState(!closed, "%s is closed", logId);
+        activeChangeSet.add(new StateChange(keyGroup, value));
+        // size threshold could be added to call persist when reached. considerations:
+        // 0. can actually degrade performance by amplifying number of requests
+        // 1. which range to persist?
+        // 2. how to deal with retries/aborts?
+    }
+
+    @Override
+    public SequenceNumber lastAppendedSequenceNumber() {
+        rollover();
+        LOG.trace("query {} sqn: {}", logId, lastAppendedSequenceNumber);
+        return lastAppendedSequenceNumber;
+    }
+
+    @Override
+    public CompletableFuture<StateChangelogHandleStreamImpl> persist(SequenceNumber from)
+            throws IOException {
+        LOG.debug("persist {} from {}", logId, from);
+        checkNotNull(from);
+        // todo: check range
+
+        rollover();
+        Collection<StoreResult> readyToReturn = confirmed.tailMap(from, true).values();
+        Collection<StateChangeSet> toUpload = changeSets.tailMap(from, true).values();
+        LOG.debug("collected readyToReturn: {}, toUpload: {}", readyToReturn, toUpload);
+        CompletableFuture<StateChangelogHandleStreamImpl> future = new CompletableFuture<>();
+        store.save(createTask(readyToReturn, toUpload, future));
+        return future;
+    }
+
+    private StoreTask createTask(
+            Collection<StoreResult> readyToReturn,
+            Collection<StateChangeSet> toUpload,
+            CompletableFuture<StateChangelogHandleStreamImpl> future) {
+        return new StoreTask(
+                toUpload,
+                future::completeExceptionally,
+                results ->
+                        executor.execute(
+                                () -> {
+                                    results.forEach(e -> uploaded.put(e.sequenceNumber, e));
+                                    future.complete(buildHandle(results, readyToReturn));
```

Comment:
Here, we need to check:
- whether a new checkpoint didn't already come in (need to resolve conflict which change to put into the `uploaded` set)
- whether the range wasn't truncated (so no need to safe, just complete the future)
- whether the checkpoint was not aborted